src/porepy/models/constitutive_laws.py

```diff
@@ -3665,23 +3689,12 @@ def thermal_stress(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
                 raise ValueError("Subdomains must be of dimension nd - 1.")
 
         discr = pp.ad.BiotAd(self.stress_keyword, subdomains)
-        alpha = self.biot_coefficient(subdomains)
-        beta = self.solid_thermal_expansion(subdomains)
-        k = self.bulk_modulus(subdomains)
-
-        # Check that both are scalar. Else, the scaling may not be correct.
-        assert isinstance(alpha, pp.ad.Scalar)
-        assert isinstance(beta, pp.ad.Scalar)
-        # The thermal stress should be multiplied by beta and k. Divide by alpha to
-        # cancel that factor from the discretization matrix.
-        stress: pp.ad.Operator = (
-            beta
-            * k
-            / alpha
-            * (
-                discr.grad_p @ self.temperature(subdomains)
-                - discr.grad_p @ self.reference_temperature(subdomains)
-            )
+        stress: pp.ad.Operator = discr.scalar_gradient(
+            self.temperature_variable
+        ) @ self.temperature(subdomains) - discr.scalar_gradient(
```

Comment:
This would equal  discr.scalar_gradient(self.temperature_variable) @ self.perturbation_from_reference(subdomains, "temperature"), wouldn't it?